### PR TITLE
feat: Add misc. repair and prepare tool methods

### PR DESCRIPTION
**Pull request by ansbdickens**
## Description
Add PrepareTools and PrepareTools.share_topology method for list of bodies
Add ExtraEdgeProblemArea.fix() method
Add RepairTools.find_short_edges and ShortEdgeProblemArea

## Issue linked
Exposes necessary components of a pyansys repair/prepare workflow

## Checklist
- [x] I have tested my changes locally.
- [x] I have added necessary documentation or updated existing documentation.
- [x] I have followed the coding style guidelines of this project.
- [x] I have added appropriate unit tests.
- [x] I have reviewed my changes before submitting this pull request.
- [x] I have linked the issue or issues that are solved to the PR if any.
- [x] I have assigned this PR to myself.
- [x] I have added the minimum version decorator to any new backend method implemented.
- [x] I have made sure that the title of my PR follows [Conventional commits style](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. ``feat: extrude circle to cylinder``)

### CODE REVIEW

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        """Initialize Prepare Tools class."""
+        self._grpc_client = grpc_client
+        self._prepare_stub = PrepareToolsStub(self._grpc_client.channel)
+        self._bodies_stub = BodiesStub(self._grpc_client.channel)
+
+    def share_topology(self, bodies: List["Body"], tol: Real = 0.0, p_instances: bool = False) -> bool:
+        """Apply ShareTopology to the chosen bodies.
+
+
+        """
```
Comment (MaxJPRey) [inline src/ansys/geometry/core/tools/prepare_tools.py:56]:
```suggestion
        """Apply ShareTopology to the chosen bodies."""
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+from ansys.geometry.core.connection import GrpcClient
+from ansys.geometry.core.typing import Real
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ansys.geometry.core.designer.body import Body
+
+
+class PrepareTools:
+    """Prepare tools for PyAnsys Geometry."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:44]:
```suggestion
    """Prepares tools for PyAnsys Geometry."""
```
Class descriptions start with a verb that ends in "s" or "es"

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    def __init__(self, grpc_client: GrpcClient):
+        """Initialize Prepare Tools class."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:47]:
```suggestion
        """Initialize ``PrepareTools`` class."""
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    def share_topology(self, bodies: List["Body"], tol: Real = 0.0, p_instances: bool = False) -> bool:
+        """Apply ShareTopology to the chosen bodies.
+
+        Parameters
+        ----------
+        bodies : List[Body]
+            List of bodies to apply ShareTopology to.
+        tol : Real
+            The maximum distance between bodies.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:60]:
```suggestion
            Maximum distance between bodies.
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    """Prepare tools for PyAnsys Geometry."""
+
+    def __init__(self, grpc_client: GrpcClient):
+        """Initialize Prepare Tools class."""
+        self._grpc_client = grpc_client
+        self._prepare_stub = PrepareToolsStub(self._grpc_client.channel)
+        self._bodies_stub = BodiesStub(self._grpc_client.channel)
+
+    def share_topology(self, bodies: List["Body"], tol: Real = 0.0, p_instances: bool = False) -> bool:
+        """Apply ShareTopology to the chosen bodies.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:53]: What is it "ShareTopology" here instead of share topology. The hyphenation in the name implies two words and it doesn't seem to be a proper noun.

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        Parameters
+        ----------
+        bodies : List[Body]
+            List of bodies to apply ShareTopology to.
+        tol : Real
+            The maximum distance between bodies.
+        p_instances : bool
+            Controls whether instances are preserved.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:62]:
```suggestion
            Whether instances are preserved.
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+        Returns
+        -------
+        bool
+            Indicates success/failure of ShareTopology operation.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/prepare_tools.py:67]:
```suggestion
            ``True`` if successful, ``False`` if failed.
```
The above is our standard statement for when a method returns a Boolean value.

File: src/ansys/geometry/core/tools/problem_areas.py [hunk 8]
```diff
@@ -268,6 +270,83 @@ def edges(self) -> List["Edge"]:
...
+        message = RepairToolMessage(
+            response.result.success,
+            response.result.created_bodies_monikers,
+            response.result.modified_bodies_monikers,
+        )
+
+        return message
+
+class ShortEdgeProblemAreas(ProblemArea):
+    """Represents a short edge problem area with unique identifier and associated edges.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/problem_areas.py:299]:
```suggestion
    """Represents a short edge problem area with a unique identifier and associated edges.
```

File: src/ansys/geometry/core/tools/problem_areas.py [hunk 9]
```diff
@@ -268,6 +270,83 @@ def edges(self) -> List["Edge"]:
...
+    id : str
+        Server-defined ID for the body.
+    grpc_client : GrpcClient
+        Active supporting geometry service instance for design modeling.
+    edges : List[Edge]
+        List of edges associated with the design.
+    """
+
+    def __init__(self, id: str, grpc_client: GrpcClient, edges: List["Edge"]):
+        """Initialize a new instance of the short edge problem area class."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/problem_areas.py:312]:
```suggestion
        """Initialize a new instance of the ``ShortEdgeProblemAreas`` class."""
```

File: src/ansys/geometry/core/tools/problem_areas.py [hunk 10]
```diff
@@ -268,6 +270,83 @@ def edges(self) -> List["Edge"]:
...
+        super().__init__(id, grpc_client)
+
+        from ansys.geometry.core.designer.edge import Edge
+
+        # Verify that all elements in the list are of type Edge
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/problem_areas.py:317]:
```suggestion
        # Verify that all elements in the are edges
```

File: src/ansys/geometry/core/tools/problem_areas.py [hunk 11]
```diff
@@ -268,6 +270,83 @@ def edges(self) -> List["Edge"]:
...
+        message = RepairToolMessage(
+            response.result.success,
+            response.result.created_bodies_monikers,
+            response.result.modified_bodies_monikers,
+        )
+
+        return message
 
 class SmallFaceProblemAreas(ProblemArea):
     """Represents a small face problem area with unique identifier and associated faces.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/problem_areas.py:352]:
```suggestion
    """Represents a small face problem area with a unique identifier and associated faces.
```

File: src/ansys/geometry/core/tools/repair_tools.py [hunk 12]
```diff
@@ -63,6 +66,7 @@ def __init__(self, grpc_client: GrpcClient):
         """Initialize Repair Tools class."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/repair_tools.py:66]:
```suggestion
        """Initialize a new instance of the ``RepairTools`` class."""
```
Try to be consistent. I rephrased the description for this class to use the same wording as the earlier ones.

File: src/ansys/geometry/core/tools/repair_tools.py [hunk 13]
```diff
@@ -177,6 +181,46 @@ def find_inexact_edges(self, bodies: List["Body"]) -> List[InexactEdgeProblemAre
...
+    ) -> List[ShortEdgeProblemAreas]:
+        """Find the short edge problem areas.
+
+        This method finds the short edge problem areas and returns a list of
+        short edge problem areas objects.
+
+        Parameters
+        ----------
+        bodies : List[Body]
+            List of bodies that short edge are investigated on.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/repair_tools.py:195]:
```suggestion
            List of bodies that short edges are investigated on.
```

File: src/ansys/geometry/core/tools/repair_tools.py [hunk 14]
```diff
@@ -177,6 +181,46 @@ def find_inexact_edges(self, bodies: List["Body"]) -> List[InexactEdgeProblemAre
             for res in problem_areas_response.result
         ]
 
+    def find_short_edges(
+            self, bodies: List["Body"], length: Real = 0.0
+    ) -> List[ShortEdgeProblemAreas]:
+        """Find the short edge problem areas.
+
+        This method finds the short edge problem areas and returns a list of
+        short edge problem areas objects.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/repair_tools.py:190]:
```suggestion
        these objects.
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    def share_topology(self, bodies: List["Body"], tol: Real = 0.0, p_instances: bool = False) -> bool:
+        """Apply ShareTopology to the chosen bodies.
+
+        Parameters
+        ----------
+        bodies : List[Body]
+            List of bodies to apply ShareTopology to.
+        tol : Real
+            The maximum distance between bodies.
+        p_instances : bool
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/prepare_tools.py:61]: Avoid shortcuts @ansbdickens
```suggestion
        preserve_instances : bool
```

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        Returns
+        -------
+        bool
+            Indicates success/failure of ShareTopology operation.
+        """
+        if not bodies:
+            return False
+
+        tolerance_value = DoubleValue(value=float(tol))
+        preserve_instances_value = BoolValue(value=p_instances)
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/prepare_tools.py:73]: Not sure you need to do this. gRPC should be able to assign the values properly if you directly pass the ``float`` or 
 ``bool`` objects

File: src/ansys/geometry/core/tools/prepare_tools.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        bodies_value = [
+            self._bodies_stub.Get(GetRequest(id=body.id, body_type=GetType.ORIGINAL)) for body in bodies
+        ]
+
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/prepare_tools.py:77]: Why is this part needed? You already have access to the bodies.

File: src/ansys/geometry/core/tools/repair_tools.py [hunk 18]
```diff
@@ -177,6 +181,46 @@ def find_inexact_edges(self, bodies: List["Body"]) -> List[InexactEdgeProblemAre
...
+        problem_areas_response = self._repair_stub.FindShortEdges(
+            FindShortEdgesRequest(
+                selection=body_ids, max_edge_length=length_value
+            )
+        )
+
+        parent_design = get_design_from_body(bodies[0])
+        return [
+            ShortEdgeProblemAreas(
+                str(res.id),
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/repair_tools.py:216]: This should already be a string AFAIK

**Review by RobPasMue — CHANGES_REQUESTED** [review]

File: src/ansys/geometry/core/tools/prepare_tools.py (hunk 17 above)
Comment (ansbdickens) [reply src/ansys/geometry/core/tools/prepare_tools.py:77]: @RobPasMue I have access to the bodies, but (to the best of my knowledge) not to all the information a Body proto message should have. In particular, I don't think that the python bodies can give me a transform_to_master matrix, and if they can fill the can_suppress, is_deleted, or is_suppressed fields, I don't know how.

File: src/ansys/geometry/core/tools/prepare_tools.py (hunk 16 above)
Comment (ansbdickens) [reply src/ansys/geometry/core/tools/prepare_tools.py:73]: They were explicitly converted in certain other methods, e.g. RepairTools.find_split_edges (see repair_tools.py, lines 96-97), so I followed that as assumed convention.

File: src/ansys/geometry/core/tools/repair_tools.py (hunk 18 above)
Comment (ansbdickens) [reply src/ansys/geometry/core/tools/repair_tools.py:216]: ShortEdgeProblemArea in models.proto transmits id as int32 (as seen [here.](https://github.com/ansys/ansys-api-geometry/blob/main/ansys/api/geometry/v0/models.proto#L268))

File: src/ansys/geometry/core/tools/prepare_tools.py (hunk 2 above)
Comment (ansbdickens) [reply src/ansys/geometry/core/tools/prepare_tools.py]: Changed to "Exposes body preparation tools"

**Comment by RobPasMue** [comment]
I will try to review this sometime today @ansbdickens - I might checkout your branch and do some modifications myself directly as well

**Comment by ansbdickens** [comment]
@RobPasMue anything I can do to help move this along?

**Comment by umutsoysalansys** [comment]
@ansbdickens are you planning to add some unit tests as well?

**Comment by RobPasMue** [comment]
I've been out on a trip until today. Let me work on this today morning - sorry for the delay. Umut brings a good point though. Tests have to be added.

**Comment by RobPasMue** [comment]
@ansbdickens - please test my changes. They should work and simplify your PR

File: doc/changelog.d/1306.added.md
Comment (RobPasMue) [inline doc/changelog.d/1306.added.md:1]: Why has this file been deleted?

File: tests/integration/test_prepare_tools.py
Comment (RobPasMue) [inline tests/integration/test_prepare_tools.py:1]: Why has this file been deleted?

**Comment by RobPasMue** [comment]
@umutsoysalansys - I am reverting your changes. They broke lots of things in the process...

**Review by RobPasMue — CHANGES_REQUESTED** [review]

**Comment by RobPasMue** [comment]
HI @umutsoysalansys  - something is wrong with this test:

```py

def test_fix_extra_edge(modeler: Modeler):
    """Test to find and fix extra edge problem areas."""
    skip_if_linux(modeler, test_fix_extra_edge.__name__, "repair_tools")  # Skip test on Linux
    design = modeler.open_file(FILES_DIR / "ExtraEdgesDesignBefore.scdocx")
    problem_areas = modeler.repair_tools.find_extra_edges(design.bodies)
    assert problem_areas[0].fix().success is True
```

Assert is throwing back False is True which means that the problem_areas are not being fixed. Can you look into it?

**Comment by umutsoysalansys** [comment]
@RobPasMue, I don't think there is anything wrong or different for this test and it is passing locally on Discovery but getting error on headless with this "Not fully implemented in Parasolid" remark. 
image

**Comment by RobPasMue** [comment]
As long as it doesn't pass on the Geometry Service @umutsoysalansys we cannot merge it. Could you solve this on the server implementation?

**Comment by RobPasMue** [comment]
Sure, we can do that - I will completely skip the test. I will open a GH issue and assign it to you to solve this problem

File: src/ansys/geometry/core/modeler.py [hunk 19]
```diff
@@ -129,6 +129,7 @@ def __init__(
             self._repair_tools = RepairTools(self._grpc_client)
             self._prepare_tools = PrepareTools(self._grpc_client)
             self._measurement_tools = MeasurementTools(self._grpc_client)
+            self._prepare_tools = PrepareTools(self._grpc_client)
```
Comment (RobPasMue) [inline src/ansys/geometry/core/modeler.py:132]:
```suggestion
```

**Review by RobPasMue — APPROVED** [review]
